Add Support for Java 21 Record Pattern : MissingSwitchDefault

child of #14890 

I have read check documentation: https://checkstyle.org/checks/coding/missingswitchdefault.html#MissingSwitchDefault
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---

```java

PS D:\CS\test> javac src/RecordPatterns.java --enable-preview --source 22
PS D:\CS\test> cat src/RecordPatterns.java
import java.util.List;

public class RecordPatterns {
    sealed interface I permits C, D { }
    final class C implements I { }
    final class D implements I { }
    record Pair<T>(T x, T y) { }

    void test() {
        C c = new C();
        D d = new D();

        Pair<I> p2 = new Pair<>(c, d);
        // exhaustiveness check by the compiler
        switch (p2) {   // violation ??
            case Pair<I>(I i, C j) -> System.out.println("Pair (C|D, C)");
            case Pair<I>(I i, D j) -> System.out.println("Pair (C|D, D)");
        }
    }

    public static void main(String[] args) {
        new RecordPatterns().test();
    }
}
PS D:\CS\test> cat config.xml

<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="MissingSwitchDefault"/>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:15:9: switch without "default" clause. [MissingSwitchDefault]
Audit done.
Checkstyle ends with 1 errors.

``` 


**Describe what you expect in detail.**

I expect no violation here since there is an exhaustiveness Check by the compiler. This means that all possible inputs must be covered. So there is no need for default

---

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java
@@ -50,9 +50,13 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * inputs must be covered.
  * </p>
  * <p>
- * See the <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.28">
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.28">
  *     Java Language Specification</a> for more information about switch statements
  *     and expressions.
+ * </p>
+ * <p>
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-14.30">
+ *     Java Language Specification</a> for more information about patterns.
  * </p>
  * <p>
  * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
@@ -116,7 +120,8 @@ public class MissingSwitchDefaultCheck extends AbstractCheck {
     }
 
     /**
-     * Checks if a switch block contains a case label with a pattern variable definition.
+     * Checks if a switch block contains a case label with a pattern variable definition
+     * or record pattern definition.
      * In this situation, the compiler enforces the given switch block to cover
      * all possible inputs, and we do not need a default label.
      *
@@ -126,7 +131,8 @@ public class MissingSwitchDefaultCheck extends AbstractCheck {
     private static boolean containsPatternCaseLabelElement(DetailAST detailAst) {
         return TokenUtil.findFirstTokenByPredicate(detailAst, ast -> {
             return ast.getFirstChild() != null
-                    && ast.getFirstChild().findFirstToken(TokenTypes.PATTERN_VARIABLE_DEF) != null;
+                    && (ast.getFirstChild().findFirstToken(TokenTypes.PATTERN_VARIABLE_DEF) != null
+                    || ast.getFirstChild().findFirstToken(TokenTypes.RECORD_PATTERN_DEF) != null);
         }).isPresent();
     }
 

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/MissingSwitchDefaultCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/MissingSwitchDefaultCheck.xml
@@ -28,9 +28,13 @@
  inputs must be covered.
  &lt;/p&gt;
  &lt;p&gt;
- See the &lt;a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.28"&gt;
+ See the &lt;a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.28"&gt;
      Java Language Specification&lt;/a&gt; for more information about switch statements
      and expressions.
+ &lt;/p&gt;
+ &lt;p&gt;
+ See the &lt;a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-14.30"&gt;
+     Java Language Specification&lt;/a&gt; for more information about patterns.
  &lt;/p&gt;</description>
          <message-keys>
             <message-key key="missing.switch.default"/>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheckTest.java
@@ -107,4 +107,12 @@ public class MissingSwitchDefaultCheckTest
                 expected);
     }
 
+    @Test
+    public void testMissingSwitchDefaultRecordPattern() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputMissingSwitchDefaultRecordPattern.java"),
+                expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefaultRecordPattern.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefaultRecordPattern.java
@@ -1,0 +1,69 @@
+/*
+MissingSwitchDefault
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.missingswitchdefault;
+
+public class InputMissingSwitchDefaultRecordPattern {
+    sealed interface I permits C, D { }
+    final class C implements I {}
+    final class D implements I { }
+    record Pair<T>(T x, T y) { }
+
+    sealed interface W {
+        record X(int x, int y) implements W { }
+        record Y(int a, int b) implements W { }
+    }
+
+    sealed interface TwoDimensional { }
+    record Point(int x, int y) implements TwoDimensional { }
+    record Line(Point start, Point end) implements TwoDimensional { }
+
+
+    void test1(Pair<I> p) {
+        switch (p) {   // ok , exhaustiveness check
+            case Pair<I>(I i, C j) -> System.out.println("Pair (C|D, C)");
+            case Pair<I>(I i, D j) -> System.out.println("Pair (C|D, D)");
+        }
+    }
+
+    void test2(W w) {
+        switch (w) {
+            case W.X(_, _) -> System.out.println("X");
+            case W.Y(_, _) -> System.out.println("Y");
+        }
+        switch (w) {
+            case W.X(int x, _) when x <= 0 -> System.out.println("X");
+            case W.X(_, _) -> System.out.println("X2");
+            case W.Y(_, _) -> System.out.println("Y");
+        }
+    }
+
+    void test3(TwoDimensional obj) {
+        switch (obj) {
+            case Point(int x, int y) : {
+                System.out.println("Point");
+                break;
+            }
+            case Line(Point start, Point end) : {
+                System.out.println("Line");
+            }
+        }
+    }
+
+    void test4(TwoDimensional obj) {
+        switch (obj) {
+            case Point(int x, int y) : {
+                System.out.println("Point");
+                break;
+            }
+            default: {
+                System.out.println("Line");
+            }
+        }
+    }
+
+}

--- a/src/xdocs/checks/coding/missingswitchdefault.xml
+++ b/src/xdocs/checks/coding/missingswitchdefault.xml
@@ -34,9 +34,14 @@
         </p>
         <p>
           See the
-          <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.28">
+          <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.28">
           Java Language Specification</a> for more information about switch statements
           and expressions.
+        </p>
+        <p>
+          See the
+          <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-14.30">
+          Java Language Specification</a> for more information about patterns.
         </p>
       </subsection>
 

--- a/src/xdocs/checks/coding/missingswitchdefault.xml.template
+++ b/src/xdocs/checks/coding/missingswitchdefault.xml.template
@@ -34,9 +34,14 @@
         </p>
         <p>
           See the
-          <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.28">
+          <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.28">
           Java Language Specification</a> for more information about switch statements
           and expressions.
+        </p>
+        <p>
+          See the
+          <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-14.30">
+          Java Language Specification</a> for more information about patterns.
         </p>
       </subsection>
 

